book/src/dev/rfcs/0006-contextual-difficulty.md

```diff
@@ -0,0 +1,813 @@
+
+#### Test network minimum difficulty implementation
+[test-net-min-difficulty-implementation]: #test-net-min-difficulty-implementation
+
+The testnet minimum difficulty calculation uses the existing
+`NetworkUpgrade::minimum_difficulty_spacing_for_height` function to calculate the
+minimum difficulty gap.
+
+In Zcash, the testnet minimum difficulty rule starts at block 299188, and in
+Zebra, contextual validation starts after Sapling activation. So we can assume
+that there is always a previous block.
+
+We implement this method on `DifficultyAdjustment`:
+```rust
+/// Returns true if the gap between the `candidate_time` and the previous block's
+/// `time` is greater than the testnet minimum difficulty time gap. The time gap
+/// depends on the `network` and `candidate_height`.
+///
+/// Returns false for `Mainnet`, when the `candidate_height` is below the testnet
+/// minimum difficulty start height, and when the time gap is too small.
+///
+/// `candidate_time` can be less than, equal to, or greater than the previous
+/// block's `time`, because block times are provided by miners.
+///
+/// Implements the testnet minimum difficulty adjustment from ZIPs 205 and 208.
+///
+/// Spec Note: Some parts of ZIPs 205 and 208 previously specified an incorrect
+/// check for the time gap. This function implements the correct "greater than"
+/// check.
+fn candidate_is_testnet_min_difficulty_block(&self) -> bool { ... }
+```
+
+`candidate_is_testnet_min_difficulty_block` is located in a new
+`zebra_chain::work::difficulty::adjustment` module.
+
+### Block difficulty threshold calculation
+[block-difficulty-threshold-calculation]: #block-difficulty-threshold-calculation
+
+The block difficulty threshold for the next block is calculated by scaling the
+mean target difficulty by the ratio between the median timespan and the averaging
+window timespan.
+
+The result of the scaled threshold calculation is limited by
+`ToCompact(PoWLimit(network))`, a per-network minimum block difficulty. This
+minimum difficulty is also used when a testnet block's time gap exceeds the
+minimum difficulty gap. We use the existing
+`ExpandedDifficulty::target_difficulty_limit` function to calculate the value of
+`ToCompact(PoWLimit(network))`.
+
+In Zebra, contextual validation starts after Sapling activation, so the genesis
+case of `Threshold()` in the Zcash specification is unreachable.
+
+Note that `zcashd` truncates the `MeanTarget` after the mean calculation, and
+after dividing by `AveragingWindowTimespan`. But as long as there is no overflow,
+this is [equivalent to the single truncation of the final result] in the Zcash
+specification. However, Zebra should follow the order of operations in `zcashd`,
+and use repeated divisions, because that can't overflow.
+
+[equivalent to the single truncation of the final result]: https://math.stackexchange.com/questions/147771/rewriting-repeated-integer-division-with-multiplication
```

Comment:
See also this clarification comment in `zcashd`:
https://github.com/zcash/zcash/pull/4860/files